pgrx-examples/arrays/src/lib.rs

```diff
@@ -100,8 +100,74 @@ fn return_vec_of_customtype() -> Vec<SomeStruct> {
     vec![SomeStruct {}]
 }
 
+#[pg_schema]
+mod vectors {
+    use pgrx::prelude::*;
+
+    extension_sql!(
+        r#"
+            do $$ begin raise warning 'creating sample data in the table ''vectors.data''.'; end; $$;
+            create table vectors.data as select vectors.random_vector(768) v from generate_series(1, 100000);
+    "#,
+        name = "vectors_data",
+        requires = [random_vector]
+    );
+
+    #[pg_extern]
+    fn sum_vector_array(input: Array<f32>) -> f32 {
+        input.iter_deny_null().map(|v| v as f32).sum()
+    }
+
+    #[pg_extern]
+    fn sum_vector_vec(input: Vec<f32>) -> f32 {
+        input.into_iter().map(|v| v as f32).sum()
+    }
+
+    #[pg_extern]
+    fn sum_vector_slice(input: Array<f32>) -> Result<f32, ArrayError> {
+        Ok(input.as_slice()?.iter().map(|v| *v as f32).sum())
+    }
+
+    #[pg_extern]
+    pub fn sum_vector_simd(values: Array<f32>) -> Result<f32, ArrayError> {
+        // this comes directly from https://stackoverflow.com/questions/23100534/how-to-sum-the-values-in-an-array-slice-or-vec-in-rust/67191480#67191480
+        // and https://www.reddit.com/r/rust/comments/13q56bp/comment/jlgq2t6/?utm_source=share&utm_medium=web2x&context=3
+        //
+        // the expectation here is that the compiler will autovectorize this code using SIMD
+        use std::convert::TryInto;
+
+        const LANES: usize = 16;
```

Comment:
This is definitely overkill (and will probably pessimize) but we don't really care here.

Reply:
I took that directly from SO.  And like a good programmer I just assume it works. 

If you’re up for improving it for clarity and maybe even perf, I’d love that. My brain isn’t yet trained to work with these concepts.

Reply:
I'll look at it when I have less on my plate/when I feel like futzing around with some optimizations.

Reply:
This idea is heavily used and tested in [arrow2](https://github.com/jorgecarleitao/arrow2/blob/main/src/compute/aggregate/sum.rs#L25) and is being used by [Polars](https://www.pola.rs/) and others for improved performance on aggregations.

We have a [repo](https://github.com/DataEngineeringLabs/simd-benches) with different implementations and corresponding benches.

Great work!

Reply:
The overkill was just 16 rather than 8 or 4, especially given that if the fold fails to be fully inlined there will be a lot of stack copies.

Reply:
(sorry, my comment was not meant to argue on that particular point!)

The background of 16 is that for a 32 bit width (e.g. f32 in SO), we have a total of 16*32=512 bits) so that the compiler can issue instructions from the [AVX-512]( https://en.wikipedia.org/wiki/AVX-512) set.

In arrow2 we have small trait system to generalize this across different types of bit widths so that the product is always 512 bits.